### PR TITLE
Improve Labbook upload image workflow (PT-184868871)

### PR DESCRIPTION
**Pull request by lublagg**

### CODE REVIEW

File: packages/labbook/src/components/runtime.tsx [hunk 1]
```diff
@@ -262,21 +312,25 @@ export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, set
...
                 authoredState={authoredState}
-                setInteractiveState={setDrawingStateFn}
+                onUploadImage={handleUploadImage}
                 disabled={disableUI}
                 onUploadStart={onUploadStart}
                 onUploadComplete={onUploadEnd}
+                text={"Upload Image"}
+                useModal={selectedItem?.data?.userBackgroundImageUrl}
+                useModalClick={handleUploadModalClick}
+                showUploadIcon={true}
```
Comment (pjanik) [inline packages/labbook/src/components/runtime.tsx:325]: This is slightly opinionated and not major, but adding these modal-related props to `UploadImage` component seems a bit strange to me, as this component doesn't really render anything related to the modal. It all happens here.

I think this could be a bit cleaner if this logic was moved to here (or even wrapped in a separate helper component if it becomes too wordy). Basically, we could render `<UploadButton>` with `onClick={handleUploadModalClick}` only if `!!selectedItem?.data?.userBackgroundImageUrl === true`. Otherwise, we'd load the `<UploadImage>` component.

**Review by pjanik — COMMENTED** [review]
I tested that in practice and it seems to work nicely, good job 👏 

I've left some inline comments. Nothing major I think, but probably some require double-checking.

File: packages/labbook/src/components/runtime.tsx [hunk 2]
```diff
@@ -193,7 +198,7 @@ export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, set
       const item = pEntries?.find(i => i.id === id);
       if (item) {
         const updatedEntry: ILabbookEntry = deepmerge(item, fields) as ILabbookEntry;
-        const newEntries = pEntries.map(i => i.id === selectedId ? updatedEntry : i);
+        const newEntries = pEntries.map(i => i === item ? updatedEntry : i);
```
[the remark this answers is not in the document]
Comment (lublagg) [reply packages/labbook/src/components/runtime.tsx:201]: Hi Piotr, this was necessary because looking at the selectedId was causing issues when we were creating a new item. When the upload is in progress in this case, we add a new item, which updates the selectedItem, but this code is being hit before that update happens, and the older selected item was being updated as well as the new one.

File: packages/labbook/src/components/runtime.tsx [hunk 3]
```diff
@@ -204,28 +209,60 @@ export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, set
...
+    }
+  };
+
+  const handleUploadImage = (url: string, mode: "replace" | "create" = "replace") => {
+    const item = mode === "create" ? addItem() : selectedItem;
+    if (item) {
+      const drawingState: IDrawingToolInteractiveState = {
+        ...(item.data || null),
+        userBackgroundImageUrl: url,
+        answerType: getAnswerType(authoredState.questionType) as any
```
[the remark this answers is not in the document]
Comment (lublagg) [reply packages/labbook/src/components/runtime.tsx:225]: This was copied over from the old code that was originally used in the UploadImage component. I can update it to use the "interactive_state" value.

File: packages/labbook/src/components/upload-image.tsx [hunk 4]
```diff
@@ -1,45 +1,49 @@
...
-export const UploadImage: React.FC<IProps> = ({ authoredState, setInteractiveState, onUploadStart, onUploadComplete, uploadButtonClass, uploadIcon, children, disabled}) => {
+export const UploadImage: React.FC<IProps> = ({ onUploadImage, uploadMode, onUploadStart, onUploadComplete, text,
+  disabled, showUploadIcon, useModal, useModalClick}) => {
   const [ uploadInProgress, setUploadInProgress ] = useState(false);
 
+  useEffect(() => {
+    return () => {
+      setUploadInProgress(false);
+    };
+  }, [setUploadInProgress]);
```
[the remark this answers is not in the document]
Comment (lublagg) [reply packages/labbook/src/components/upload-image.tsx:37]: Yes. Previously we were getting this warning in the console:

Warning: Can't perform a React state update on an unmounted component.
This is a no-op, but it indicates a memory leak in your application.
To fix, cancel all subscriptions and asynchronous tasks in a useEffect
cleanup function.

This stopped the warning.

File: packages/labbook/src/components/runtime.tsx (hunk 1 above)
Comment (lublagg) [reply packages/labbook/src/components/runtime.tsx:325]: Thanks Piotr, yeah I originally had it that way but thought it was a bit too wordy / confusing. But I can change it back.

File: packages/labbook/src/components/upload-image.tsx [hunk 5]
```diff
@@ -1,45 +1,49 @@
...
+  onUploadImage: (url: string, mode?: "replace" | "create") => void;
   onUploadStart?: () => void;
   onUploadComplete?: (result: { success: boolean }) => void;
-  uploadButtonClass?: 'string';
-  uploadIcon?: React.ReactNode;
   disabled?: boolean;
+  text?: string;
+  showUploadIcon?: boolean;
+  useModal?: string;
+  useModalClick?: () => void;
```
[the remark this answers is not in the document]
Comment (lublagg) [reply packages/labbook/src/components/upload-image.tsx:25]: Will do.

File: packages/labbook/src/components/upload-image.tsx [hunk 6]
```diff
@@ -1,45 +1,49 @@
...
-  setInteractiveState?: (updateFunc: (prevState: IGenericInteractiveState | null) => IGenericInteractiveState) => void;
+  onUploadImage: (url: string, mode?: "replace" | "create") => void;
   onUploadStart?: () => void;
   onUploadComplete?: (result: { success: boolean }) => void;
-  uploadButtonClass?: 'string';
-  uploadIcon?: React.ReactNode;
   disabled?: boolean;
+  text?: string;
+  showUploadIcon?: boolean;
+  useModal?: string;
```
[the remark this answers is not in the document]
Comment (lublagg) [reply packages/labbook/src/components/upload-image.tsx:24]: Will change.

File: packages/labbook/src/components/upload-image.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (pjanik) [reply packages/labbook/src/components/upload-image.tsx:34]: Ah, this was happening when user closed the dialog during the upload I guess? Okay.

File: packages/labbook/src/components/runtime.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (pjanik) [reply packages/labbook/src/components/runtime.tsx]: I see. It had to use the helper as UploadImage is a shared component and is used by Labbook, Drawing Tool, and Image Question (that has a different answer type). Here's it's gonna be always "interactive_state". Nothing major, but can be a bit cleaner.

File: packages/labbook/src/components/runtime.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (pjanik) [reply packages/labbook/src/components/runtime.tsx:202]: I see.

> As for the second point about checking id's, I though because of how item is defined on line 198, we could just use the item that way. But let me know if you think I should change it.

That's true, it cannot fail currently because of the line 198 and the fact that nothing happens to item between. 👍 

However, if we have access to any kind of `id`, it can be usually safer to use that. Just like a rule of thumb. Imagine someone changes line 198 to:
`const item = { ...pEntries?.find(i => i.id === id), oldItem: true };`
Ofc it doesn't make sense here, but just an example. That's enough to break comparisons relying on reference.

**Review by pjanik — APPROVED** [review]
Looks good 👍
